homework/hw_w02.py

```diff
@@ -0,0 +1,253 @@
+    date_df['year'] = pd.to_datetime(date_df.iso_date).dt.year
+    date_df['quarter'] = pd.to_datetime(date_df.iso_date).dt.quarter
+    date_df['month'] = pd.to_datetime(date_df.iso_date).dt.month
+    date_df['week'] = pd.to_datetime(date_df.iso_date).dt.week
+    date_df['day_of_year'] = pd.to_datetime(date_df.iso_date).dt.dayofyear
+    date_df['day_of_month'] = pd.to_datetime(date_df.iso_date).dt.day
+    dayOfWeek={0:'Monday', 1:'Tuesday', 2:'Wednesday', 3:'Thursday',
+        4:'Friday', 5:'Saturday', 6:'Sunday'}
+    date_df['day_of_week'] = (pd.to_datetime(date_df.iso_date)
+        .dt.dayofweek.map(dayOfWeek))
+    # This converts the date to a string in ISO format, YYYY-MM-DD
+    # and converting the NaT datetime to 0000-00-00 to avoid having
+    # to drop those columns
+    date_df['iso_date'] = (pd.to_datetime(date_df['iso_date'])
+                               .apply(lambda x: x.strftime('%Y-%m-%d')
+                                      if not pd.isnull(x) else '0000-00-00'))
+    # Fix all of the other NaTs to None
+    date_df = date_df.where(pd.notnull(date_df), None)
+    return date_df
+
+def etl_measures_table(join_query, conn_op_db, conn_star_db,
+        table='fact_order_items'):
+    measures_df = create_df(join_query, conn_op_db)
+    # Convert the datetimes to strings for easier use as keys
+    fix_measures_dt(measures_df)
+    # Create derived measures for use in summaries
+    create_derived_measures(measures_df)
+    load_table(conn_star_db, table, measures_df)
+
+def fix_measures_dt(measures_df):
+    for key in ["order_date",
+                "required_date",
+                "shipped_date",]:
+        # Similar to the correction done on the dates, convert the dates
+        # to strings for use as keys, setting NaT to 0000-00-00
+        measures_df[key] = (pd.to_datetime(measures_df[key])
+                           .apply(lambda x: x.strftime('%Y-%m-%d')
+                                  if not pd.isnull(x) else '0000-00-00'))
+
+def create_derived_measures(measures_df):
+    # Transforms
+    measures_df['revenue'] = (measures_df['quantity_ordered'] *
+            measures_df['price_each'])
+    measures_df['costs'] = (measures_df['quantity_ordered'] *
+            measures_df['buy_price'])
+    measures_df['profits'] = (measures_df['revenue'] -
+            measures_df['costs'])
+    measures_df['profit_per_item'] = (measures_df['profits'] /
+            measures_df['quantity_ordered'])
+    measures_df['profit_margin'] = (measures_df['profits'] /
+            measures_df['revenue'])
+
+###-###-###-###-###-###-###-###-###-###-###-###-###-###-###-###-###-###-###-
+# Run script
+###-###-###-###-###-###-###-###-###-###-###-###-###-###-###-###-###-###-###-
+if __name__ == "__main__":
+    conn_op_db = DBConnection.DBConnection(dbname="company_db")
+    conn_star_db = DBConnection.DBConnection(dbname="star_db")
+
+    # Using an ordered dictionary as the measures table must be loaded last
```

Comment:
Why not just use a list of tuples??? Ordered dict is for when you need the order AND the O(1) lookup by key...